info-articles/Mapping in solidity - issue #19.md/article.md

```diff
@@ -0,0 +1,437 @@
+ 
+      function getTheLast() public view returns(uint) {
+       return theBalanceOf[keys[keys.length -1]];
+      }
+```
+These two getter functions retrieved the balances of key[0], which is the array's first element, and the array's last member, which is the length minus 1.
+
+## Creating a Custom Contract in Solidity to Explain Mapping
+Since we have learned much about mapping, let us create a simple contract and use the knowledge in real time.
+```
+// SPDX-License-Identifier: MIT
+ pragma solidity ^0.8.16;
+ 
+  contract TheWallet{
+      address public John;
+ 
+      constructor() {
+          John == msg.sender;
+      }
+   
+    mapping(address => uint) public hisBalance;
+    mapping (address=> mapping(address => uint)) public allowThem;
+   
+    function deposit() public payable {
+        hisBalance[msg.sender] += msg.value;
+    }
+ 
+    function withdrawal() public payable {
+        (bool sent, ) = msg.sender.call{value: hisBalance[msg.sender]}("");
+        require(sent, "cannot withdraw");
+    }
+ 
+    function transfer(address receiver, uint token) public returns(bool){
+        allowThem[msg.sender][receiver] += token;
+        hisBalance[msg.sender] += token;
+        hisBalance[receiver] -= token;
+        return true;
+    }
+ 
+    function remainingMoney() public view returns (uint) {
+        return address(this).balance;
+    }
+  }
+ ```
+ 
+
+This contract is a simple wallet contract where users can deposit, transfer, and know the remaining Ethers in the contract. Here, we used mapping to track both balances and allowance.
+
+The next stage is to compile and deploy this contract using Hardhat. Go to the Hardhat folder on your device, locate the src folder and write your contract there. Perhaps you have not installed Hardhat in your terminal, paste this on your Vs Code terminal:
+```
+npm install — save-dev hardhat
+```
+Then paste this:
+```
+npm install --save-dev "@nomiclabs/hardhat-ethers@^2.0.0"
+```
+For the compilation, run this command:
+```
+npx hardhat compile
+```
+Once that is successful, the next phase is deployment to your favorite testnet using the Chainstack Ethereum URL. First, you will need to configure your hardhat. Therefore, locate a hardhat.config file and paste this script. You can create a new one if you could not find it too.
+```
+require("@nomicfoundation/hardhat-toolbox");
+
+const Chainstack_API_KEY = "KEY";
+
+const Mumbai_PRIVATE_KEY = "YOUR GOERLI PRIVATE KEY";
+module.exports = {
+  solidity: "0.8.17",
+
+ networks: {
+
+   goerli: {
+
+     url: `https://eth-goerli.chainstackapi.io/v2/${Chainstack_API_KEY}`,
+          // simply add your Chainstack URL here
+
+     accounts: [GOERLI_PRIVATE_KEY]
+   }
+
+ }
+
+};
+```
+The final step is to write a deployment script. Create a deploy.js file where you will write your deployment script - you can choose either Javascript or Typescript. When you are done, start a node:
+```
+npx hardhat node
+```
+Then you can deploy to a testnet:

```

Comment:
First a deploy script specific to the contract needs to be written to deploy a contract before calling the script.

Reply:
Indicated that.